ldb_reader.go

```diff
@@ -244,14 +313,30 @@ func (reader *LDBReader) Close() error {
 	reader.mu.Lock()
 	defer reader.mu.Unlock()
 
+	if reader.cancelWatcher != nil {
+		reader.cancelWatcher()
+	}
+
+	return reader.closeDB()
+}
+
+// closeDB closes all reader-owned resources associated with the current DB.
+// It should only be called when the caller is holding the reader.mu mutex.
+func (reader *LDBReader) closeDB() error {
 	for _, stmt := range reader.getRowByKeyStmtCache {
 		stmt.Close()
 	}
+	reader.getRowByKeyStmtCache = map[string]*sql.Stmt{}
```

Comment:
Why are we clearing these caches here?  Are we planning on re-using the LDBReader after closing?

edit: reading further it looks like we're re-using it when switching versions. i'd prefer a higher level abstraction that loaned out a valid LDB reader to the caller and will also handle the shutting down of previous LDBs that have been replaced by a newer version.

the reason is because over time depending on being able to clean up internal state when switching between versions is likely to cause issues when cleanup steps are missed as other state is added to the reader. the common pattern of discarding `Close()`'d data would be safer here.

Could we have something like

```
type VersionedLDBReader struct { ... }

vLDB := newVersionedLDBReader(basePath)
vLDB.Loan(func (r *LDBReader) { .... }) 
```

?

Reply:
I'd prefer we abstract away how an LDB is provided from the external ctlstore API. I don't think we should leak to clients that an LDB is provided via LDB versioning instead of ledger reflection. But, CMIIW and misunderstanding what you're advocating for :)

Reply:
I think some addition types like the loaner (which can just be a `func() *LDBReader` in the client code) would make a lot of sense but if you want to move forward with this approach feel free to.

Reply:
I'd prefer as-is :) Appreciate the alternative feedback though!